### PR TITLE
[BREAKING] Remove first parameter of `Element` constructor

### LINKED ISSUE #260: Remove first parameter of `Element` constructor
Currently, the [`Element` constructor](https://playcanvas.github.io/pcui/typedocs/classes/Element.html#constructor) takes 2 arguments:

* A DOM element (or string representing a DOM element that can be passed to `document.createElement`).
* An options object. The options object allows for a `dom` property to be set which duplicates what can be specified by the first argument.

The `Element` constructor signature differs from all of its subclasses in having this first element. I propose we:

* Drop this first argument and instead rely solely on the `dom` property of the options object (which defaults to `div` if not specified).
* Bump major version number to 4.0.0 since this is a breaking API change.

### PR DESCRIPTION
**Pull request by willeastcott**
At the moment, you can set the DOM element when creating a PCUI Element by setting the first parameter of the `Element` constructor or by setting the `dom` property of the `ElementArgs` options object. The code gives precedence to the first parameter of the constructor, but otherwise, there appears to be little reason why the parameter is needed at all. It makes the constructor signature of `Element` needlessly different to every other PCUI element class. Therefore, this PR removes it.

As an aside, this PR makes all constructor options objects read-only. The utility type `Readonly<...>` has been applied to the options object parameter types. For instances where an options objects must to be written to, it is shallow copied to a new object using object destructuring.

Fixes #260

### CODE REVIEW

File: src/components/BooleanInput/index.ts [hunk 1]
```diff
@@ -27,8 +27,9 @@ class BooleanInput extends Input implements IBindable, IFocusable {
     protected _value: boolean;
 
     constructor(args: BooleanInputArgs = {}) {
-        args.tabIndex = args.tabIndex ?? 0;
-        super(args.dom, args);
+        args.tabIndex ??= 0;
```
Comment (slimbuck) [inline src/components/BooleanInput/index.ts:30]: Wondering whether destructuring directly and leaving the original args might be better?

Comment (willeastcott) [reply src/components/BooleanInput/index.ts:30]: I was wondering the same thing myself. But in the case of `tabIndex`, the code was updating the supplied options object before this PR. This PR does not also update the supplied options object's dom property in some places now though. Hmm, I guess technically that's not great....

Comment (willeastcott) [reply src/components/BooleanInput/index.ts:30]: So you'd prefer this?

```
        if (args.tabIndex === undefined) {
            args = {
                ...args,
                tabIndex: 0
            };
        }
```

Comment (slimbuck) [reply src/components/BooleanInput/index.ts:30]: Either that or just:
`super({ tabIndex: 0, …tags })`

Comment (willeastcott) [reply src/components/BooleanInput/index.ts:30]: But the user might have set `tabIndex` themselves. Your suggestion would always ignore the user-supplied value, no?

Comment (willeastcott) [reply src/components/BooleanInput/index.ts]: I've updated the code. There should no longer be any instances where the supplied options object is modified.

Comment (slimbuck) [reply src/components/BooleanInput/index.ts]: Doing:
```
super({ tabindex: 0, ...args });
```
will result in `tabIndex` being overridden by args if it has it present.

File: src/components/ArrayInput/index.ts [hunk 2]
```diff
@@ -117,17 +117,17 @@ class ArrayInput extends Element implements IFocusable, IBindable {
...
-        delete args.binding;
-
+    constructor(args: Readonly<ArrayInputArgs> = {}) {
         const container = new Container({
             dom: args.dom,
             flex: true
         });
 
-        super(container.dom, args);
+        const elementArgs = { ...args, dom: container.dom };
```
Comment (slimbuck) [inline src/components/ArrayInput/index.ts:126]: Think you want the order of `...args` and `dom:` swapped.

Comment (slimbuck) [reply src/components/ArrayInput/index.ts:126]: Unless I'm misunderstanding things!

Comment (slimbuck) [reply src/components/ArrayInput/index.ts:126]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#spread_in_object_literals

Comment (willeastcott) [reply src/components/ArrayInput/index.ts:126]: The code before forced the dom arg to the `container.dom` value. So we want to override the args object with the `container.dom` too. If the values were reversed, the args could potentially replace `container.dom`.

File: src/components/BooleanInput/index.ts (hunk 1 above)
Comment (willeastcott) [reply src/components/BooleanInput/index.ts]: Oh yeah, cool! So I've updated the code again to use this where applicable. Thanks!

File: src/components/ArrayInput/index.ts (hunk 2 above)
Comment (slimbuck) [reply src/components/ArrayInput/index.ts:126]: Oic that's expected! 👍

**Review by slimbuck — APPROVED** [review]

**Review by ellthompson — APPROVED** [review]

**Pull request merged**